File: binding/include/multimedia/libqwavedecoder.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qwavedecoder.html
pub const qwavedecoder = struct {

    /// New constructs a new QWaveDecoder object.
    pub fn New(device: ?*anyopaque) QtC.QWaveDecoder {
        return qtc.QWaveDecoder_new(@ptrCast(device));
    }


    /// New2 constructs a new QWaveDecoder object.
    pub fn New2(device: ?*anyopaque, format: ?*anyopaque) QtC.QWaveDecoder {
        return qtc.QWaveDecoder_new2(@ptrCast(device), @ptrCast(format));
    }


    /// New3 constructs a new QWaveDecoder object.
    pub fn New3(device: ?*anyopaque, parent: ?*anyopaque) QtC.QWaveDecoder {
        return qtc.QWaveDecoder_new3(@ptrCast(device), @ptrCast(parent));
    }


    /// New4 constructs a new QWaveDecoder object.
    pub fn New4(device: ?*anyopaque, format: ?*anyopaque, parent: ?*anyopaque) QtC.QWaveDecoder {
        return qtc.QWaveDecoder_new4(@ptrCast(device), @ptrCast(format), @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWaveDecoder_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qwavedecoder.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn AudioFormat(self: ?*anyopaque, ) QtC.QAudioFormat {
        return qtc.QWaveDecoder_AudioFormat(@ptrCast(self));
    }

    pub fn GetDevice(self: ?*anyopaque, ) QtC.QIODevice {
        return qtc.QWaveDecoder_GetDevice(@ptrCast(self));
    }

    pub fn Duration(self: ?*anyopaque, ) i32 {
        return qtc.QWaveDecoder_Duration(@ptrCast(self));
    }

    pub fn HeaderLength() i64 {
        return qtc.QWaveDecoder_HeaderLength();
    }

    pub fn Open(self: ?*anyopaque, mode: QIODevice::OpenMode) bool {
        return qtc.QWaveDecoder_Open(@ptrCast(self), mode);
    }

    /// Allows for overriding the related default method
    pub fn OnOpen(self: ?*anyopaque, callback: *const fn (?*anyopaque, QIODevice::OpenMode) callconv(.c) bool) void {
        qtc.QWaveDecoder_OnOpen(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseOpen(self: ?*anyopaque, mode: QIODevice::OpenMode) bool {
        return qtc.QWaveDecoder_QBaseOpen(@ptrCast(self), mode);
    }

    pub fn Close(self: ?*anyopaque, ) void {
        qtc.QWaveDecoder_Close(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnClose(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QWaveDecoder_OnClose(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseClose(self: ?*anyopaque, ) void {
        qtc.QWaveDecoder_QBaseClose(@ptrCast(self));
    }

    pub fn Seek(self: ?*anyopaque, pos: i64) bool {
        return qtc.QWaveDecoder_Seek(@ptrCast(self), pos);
    }

    /// Allows for overriding the related default method
    pub fn OnSeek(self: ?*anyopaque, callback: *const fn (?*anyopaque, i64) callconv(.c) bool) void {
        qtc.QWaveDecoder_OnSeek(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSeek(self: ?*anyopaque, pos: i64) bool {
        return qtc.QWaveDecoder_QBaseSeek(@ptrCast(self), pos);
    }

    pub fn Pos(self: ?*anyopaque, ) i64 {
        return qtc.QWaveDecoder_Pos(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnPos(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i64) void {
        qtc.QWaveDecoder_OnPos(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBasePos(self: ?*anyopaque, ) i64 {
        return qtc.QWaveDecoder_QBasePos(@ptrCast(self));
    }

    pub fn SetIODevice(self: ?*anyopaque, device: ?*anyopaque) void {
        qtc.QWaveDecoder_SetIODevice(@ptrCast(self), @ptrCast(device));
    }

    pub fn Size(self: ?*anyopaque, ) i64 {
        return qtc.QWaveDecoder_Size(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnSize(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i64) void {
        qtc.QWaveDecoder_OnSize(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSize(self: ?*anyopaque, ) i64 {
        return qtc.QWaveDecoder_QBaseSize(@ptrCast(self));
    }

    pub fn IsSequential(self: ?*anyopaque, ) bool {
        return qtc.QWaveDecoder_IsSequential(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnIsSequential(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) bool) void {
        qtc.QWaveDecoder_OnIsSequential(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseIsSequential(self: ?*anyopaque, ) bool {
        return qtc.QWaveDecoder_QBaseIsSequential(@ptrCast(self));
    }

    pub fn BytesAvailable(self: ?*anyopaque, ) i64 {
        return qtc.QWaveDecoder_BytesAvailable(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnBytesAvailable(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i64) void {
        qtc.QWaveDecoder_OnBytesAvailable(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseBytesAvailable(self: ?*anyopaque, ) i64 {
        return qtc.QWaveDecoder_QBaseBytesAvailable(@ptrCast(self));
    }

    pub fn FormatKnown(self: ?*anyopaque, ) void {
        qtc.QWaveDecoder_FormatKnown(@ptrCast(self));
    }

    pub fn OnFormatKnown(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QWaveDecoder_Connect_FormatKnown(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn ParsingError(self: ?*anyopaque, ) void {
        qtc.QWaveDecoder_ParsingError(@ptrCast(self));
    }

    pub fn OnParsingError(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QWaveDecoder_Connect_ParsingError(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWaveDecoder_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qwavedecoder.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWaveDecoder_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qwavedecoder.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QWaveDecoder_Delete(@ptrCast(self));
    }
};

File: binding/src/multimedia/libqwavedecoder.cpp
#include <QAudioFormat>
#include <QIODevice>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QWaveDecoder>
#include <qwavedecoder.h>
#include "libqwavedecoder.h"
#include "libqwavedecoder.hxx"

QWaveDecoder* QWaveDecoder_new(QIODevice* device) {
	 return new VirtualQWaveDecoder(device);
}

QWaveDecoder* QWaveDecoder_new2(QIODevice* device, const QAudioFormat* format) {
	 return new VirtualQWaveDecoder(device, *format);
}

QWaveDecoder* QWaveDecoder_new3(QIODevice* device, QObject* parent) {
	 return new VirtualQWaveDecoder(device, parent);
}

QWaveDecoder* QWaveDecoder_new4(QIODevice* device, const QAudioFormat* format, QObject* parent) {
	 return new VirtualQWaveDecoder(device, *format, parent);
}

libqt_string QWaveDecoder_Tr(const char* s) {
	QString _ret = QWaveDecoder::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QAudioFormat* QWaveDecoder_AudioFormat(const QWaveDecoder* self) {
	return new QAudioFormat(self->audioFormat());
}

QIODevice* QWaveDecoder_GetDevice(QWaveDecoder* self) {
	return self->getDevice();
}

int QWaveDecoder_Duration(const QWaveDecoder* self) {
	return self->duration();
}

long long QWaveDecoder_HeaderLength() {
	return QWaveDecoder::headerLength();
}

bool QWaveDecoder_Open(QWaveDecoder* self, QIODevice::OpenMode mode) {
	return self->open(mode);
}

void QWaveDecoder_Close(QWaveDecoder* self) {
	self->close();
}

bool QWaveDecoder_Seek(QWaveDecoder* self, long long pos) {
	return self->seek(pos);
}

long long QWaveDecoder_Pos(const QWaveDecoder* self) {
	return self->pos();
}

void QWaveDecoder_SetIODevice(QWaveDecoder* self, QIODevice* device) {
	self->setIODevice(device);
}

long long QWaveDecoder_Size(const QWaveDecoder* self) {
	return self->size();
}

bool QWaveDecoder_IsSequential(const QWaveDecoder* self) {
	return self->isSequential();
}

long long QWaveDecoder_BytesAvailable(const QWaveDecoder* self) {
	return self->bytesAvailable();
}

void QWaveDecoder_FormatKnown(QWaveDecoder* self) {
	self->formatKnown();
}

void QWaveDecoder_Connect_FormatKnown(QWaveDecoder* self, intptr_t slot) {
    void (*slotFunc)(QWaveDecoder*) = reinterpret_cast<void (*)(QWaveDecoder*)>(slot);
    QWaveDecoder::connect(self, &QWaveDecoder::formatKnown, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QWaveDecoder_ParsingError(QWaveDecoder* self) {
	self->parsingError();
}

void QWaveDecoder_Connect_ParsingError(QWaveDecoder* self, intptr_t slot) {
    void (*slotFunc)(QWaveDecoder*) = reinterpret_cast<void (*)(QWaveDecoder*)>(slot);
    QWaveDecoder::connect(self, &QWaveDecoder::parsingError, [self, slotFunc]() {
	slotFunc(self);
    });
}

libqt_string QWaveDecoder_Tr2(const char* s, const char* c) {
	QString _ret = QWaveDecoder::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QWaveDecoder_Tr3(const char* s, const char* c, int n) {
	QString _ret = QWaveDecoder::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
bool QWaveDecoder_QBaseOpen(QWaveDecoder* self, QIODevice::OpenMode mode) {
	auto* vqwavedecoder = dynamic_cast<VirtualQWaveDecoder*>(self);
	if (vqwavedecoder && vqwavedecoder->isVirtualQWaveDecoder) {
vqwavedecoder->setQWaveDecoder_Open_IsBase(true);
	return vqwavedecoder->open(mode);
}
}

// Auxiliary method to allow providing re-implementation
void QWaveDecoder_OnOpen(QWaveDecoder* self, intptr_t slot) {
	auto* vqwavedecoder = dynamic_cast<VirtualQWaveDecoder*>(self);
	if (vqwavedecoder && vqwavedecoder->isVirtualQWaveDecoder) {
vqwavedecoder->setQWaveDecoder_Open_Callback(reinterpret_cast<VirtualQWaveDecoder::QWaveDecoder_Open_Callback>(slot));
}
}

// Base class handler implementation
void QWaveDecoder_QBaseClose(QWaveDecoder* self) {
	auto* vqwavedecoder = dynamic_cast<VirtualQWaveDecoder*>(self);
	if (vqwavedecoder && vqwavedecoder->isVirtualQWaveDecoder) {
vqwavedecoder->setQWaveDecoder_Close_IsBase(true);
	vqwavedecoder->close();
}
}

// Auxiliary method to allow providing re-implementation
void QWaveDecoder_OnClose(QWaveDecoder* self, intptr_t slot) {
	auto* vqwavedecoder = dynamic_cast<VirtualQWaveDecoder*>(self);
	if (vqwavedecoder && vqwavedecoder->isVirtualQWaveDecoder) {
vqwavedecoder->setQWaveDecoder_Close_Callback(reinterpret_cast<VirtualQWaveDecoder::QWaveDecoder_Close_Callback>(slot));
}
}

// Base class handler implementation
bool QWaveDecoder_QBaseSeek(QWaveDecoder* self, long long pos) {
	auto* vqwavedecoder = dynamic_cast<VirtualQWaveDecoder*>(self);
	if (vqwavedecoder && vqwavedecoder->isVirtualQWaveDecoder) {
vqwavedecoder->setQWaveDecoder_Seek_IsBase(true);
	return vqwavedecoder->seek(pos);
}
}

// Auxiliary method to allow providing re-implementation
void QWaveDecoder_OnSeek(QWaveDecoder* self, intptr_t slot) {
	auto* vqwavedecoder = dynamic_cast<VirtualQWaveDecoder*>(self);
	if (vqwavedecoder && vqwavedecoder->isVirtualQWaveDecoder) {
vqwavedecoder->setQWaveDecoder_Seek_Callback(reinterpret_cast<VirtualQWaveDecoder::QWaveDecoder_Seek_Callback>(slot));
}
}

// Base class handler implementation
long long QWaveDecoder_QBasePos(const QWaveDecoder* self) {
	auto* vqwavedecoder = dynamic_cast<const VirtualQWaveDecoder*>(self);
	if (vqwavedecoder && vqwavedecoder->isVirtualQWaveDecoder) {
vqwavedecoder->setQWaveDecoder_Pos_IsBase(true);
	return vqwavedecoder->pos();
}
}

// Auxiliary method to allow providing re-implementation
void QWaveDecoder_OnPos(const QWaveDecoder* self, intptr_t slot) {
	auto* vqwavedecoder = dynamic_cast<const VirtualQWaveDecoder*>(self);
	if (vqwavedecoder && vqwavedecoder->isVirtualQWaveDecoder) {
vqwavedecoder->setQWaveDecoder_Pos_Callback(reinterpret_cast<VirtualQWaveDecoder::QWaveDecoder_Pos_Callback>(slot));
}
}

// Base class handler implementation
long long QWaveDecoder_QBaseSize(const QWaveDecoder* self) {
	auto* vqwavedecoder = dynamic_cast<const VirtualQWaveDecoder*>(self);
	if (vqwavedecoder && vqwavedecoder->isVirtualQWaveDecoder) {
vqwavedecoder->setQWaveDecoder_Size_IsBase(true);
	return vqwavedecoder->size();
}
}

// Auxiliary method to allow providing re-implementation
void QWaveDecoder_OnSize(const QWaveDecoder* self, intptr_t slot) {
	auto* vqwavedecoder = dynamic_cast<const VirtualQWaveDecoder*>(self);
	if (vqwavedecoder && vqwavedecoder->isVirtualQWaveDecoder) {
vqwavedecoder->setQWaveDecoder_Size_Callback(reinterpret_cast<VirtualQWaveDecoder::QWaveDecoder_Size_Callback>(slot));
}
}

// Base class handler implementation
bool QWaveDecoder_QBaseIsSequential(const QWaveDecoder* self) {
	auto* vqwavedecoder = dynamic_cast<const VirtualQWaveDecoder*>(self);
	if (vqwavedecoder && vqwavedecoder->isVirtualQWaveDecoder) {
vqwavedecoder->setQWaveDecoder_IsSequential_IsBase(true);
	return vqwavedecoder->isSequential();
}
}

// Auxiliary method to allow providing re-implementation
void QWaveDecoder_OnIsSequential(const QWaveDecoder* self, intptr_t slot) {
	auto* vqwavedecoder = dynamic_cast<const VirtualQWaveDecoder*>(self);
	if (vqwavedecoder && vqwavedecoder->isVirtualQWaveDecoder) {
vqwavedecoder->setQWaveDecoder_IsSequential_Callback(reinterpret_cast<VirtualQWaveDecoder::QWaveDecoder_IsSequential_Callback>(slot));
}
}

// Base class handler implementation
long long QWaveDecoder_QBaseBytesAvailable(const QWaveDecoder* self) {
	auto* vqwavedecoder = dynamic_cast<const VirtualQWaveDecoder*>(self);
	if (vqwavedecoder && vqwavedecoder->isVirtualQWaveDecoder) {
vqwavedecoder->setQWaveDecoder_BytesAvailable_IsBase(true);
	return vqwavedecoder->bytesAvailable();
}
}

// Auxiliary method to allow providing re-implementation
void QWaveDecoder_OnBytesAvailable(const QWaveDecoder* self, intptr_t slot) {
	auto* vqwavedecoder = dynamic_cast<const VirtualQWaveDecoder*>(self);
	if (vqwavedecoder && vqwavedecoder->isVirtualQWaveDecoder) {
vqwavedecoder->setQWaveDecoder_BytesAvailable_Callback(reinterpret_cast<VirtualQWaveDecoder::QWaveDecoder_BytesAvailable_Callback>(slot));
}
}

void QWaveDecoder_Delete(QWaveDecoder* self) {
    delete self;
}

